### DISCUSSION TITLE
Propagated OpenTelemetry Trace Information to downstream KubernetesPodOperator jobs

### DISCUSSION THREAD

**Discussion by ranimufid** [apache/airflow#43352] [Q&A]
Airflow Version: `2.10.1`
Airflow managed by AWS on MWAA 

## Goal
Hi folks. I'm trying to get _custom_ opentelemetry instrumentation to work with DAGs that have been dispatched as K8s pods using the `KubernetesPodOperator`. And to be specific, I am trying to enrich the OTEL trace information with custom span names and maybe even emit some custom metrics from individual DAG runs in kubernetes.

## Current State
✅  I have successfully enabled OTEL on Airflow (MWAA) and i'm able to see spans and traces originating from Airflow core components.

![Screenshot 2024-10-24 at 16 35 07](https://github.com/user-attachments/assets/e1291591-a83d-4864-966c-024951047723)

⚠️ However, none of the custom instrumentation (custom span names and traces) that i'm adding inside the DAG code get propagated to the OTEL collector.

### DAG Dispatch
```python
from airflow.contrib.operators.kubernetes_pod_operator import KubernetesPodOperator

from global_dag_defaults import (
    K8S_PYTHON_TASK_KWARGS,
)

otel_demo = KubernetesPodOperator(
    cmds=[
        "python3",
        "/dags/otel_demo.py",
    ],
    name="otel_demo",
    task_id="otel_demo",
    dag=dag,
    **K8S_PYTHON_TASK_KWARGS
)

otel_demo
```

### DAG Code: `otel_demo.py`

```python
import requests
from opentelemetry import metrics, trace

tracer = trace.get_tracer(__name__)
meter = metrics.get_meter(__name__)

my_custom_counter = meter.create_counter(
    "my_custom_counter",
    description="Is incremented on X",
    unit="1",
)

@tracer.start_as_current_span("call_github")
def call_github():
    my_custom_counter.add(1, {"custom_attribute": "custom_value"})
    with tracer.start_as_current_span(name="Make GitHub call"):
        r = requests.get('https://api.github.com/events')

```

## Observations

- The DAG run logs show the following log line which created the impression that the trace context was being propagated to the kubernetes pods running the DAG code, but that doesn't seem to be the case
```
[2024-10-23, 18:03:24 UTC] {otel_tracer.py:265} INFO - [OTLPSpanExporter] Connecting to OpenTelemetry Collector at https://placeholder.collector.dns:443/v1/traces
```

- To confirm that the custom instrumentation works, I added a OTEL collector as a side car to the airflow dispatched pod and I was able to receive the spans and traces

❓ So the ultimate question is: how can one propagate trace context and the collector settings to airflow jobs dispatched as k8s pods?

**Comment by potiuk** [comment]
I think (@howardyoo ?) ->  you can't now, but it woould be a nice feature to add.

**Reply by howardyoo** [reply]
Yes, to add that, I would even say not only to the pod operators, but operators in general it would be great if there's a common way to make the trace context available in a sort of standardized and transparent way would be great. This could actually be a good project on its own.

**Comment by mxmrlt** [comment]
Hi.

For now you can manage to do it thanks to @howardyoo OTEL Provider https://github.com/howardyoo/airflow_otel_provider and AOP programming.

Sure it's not the perfect solution but it works. That's how I can monitor my Airflow instance within Elastic APM.
Of course I'd like it to be a native feature.

Here's an example with Kafka Operator.

```python
@aspectlib.Aspect
def instrument_producer(*args, **kwargs):
    producer = yield aspectlib.Proceed
    instrumentation = ConfluentKafkaInstrumentor()
    producer = instrumentation.instrument_producer(
        producer=producer, tracer_provider=otel_hook.tracer_provider
    )

    yield aspectlib.Return(producer)

@aspectlib.Aspect(bind=True)
def instrument_execute(cutpoint, *args, **kwargs):

    from airflow.operators.python import get_current_context

    dag_context = get_current_context()

    with otel_hook.start_as_current_span(
        name=cutpoint.__name__, dag_context=dag_context
    ) as s:
        yield aspectlib.Proceed(*args, **kwargs)


aspectlib.weave(KafkaProducerHook.get_producer, instrument_producer)
aspectlib.weave(target=ProduceToTopicOperator.execute, aspects=instrument_execute)

produce_treats = ProduceToTopicOperator(
    task_id="produce_treats",
    kafka_config_id="kafka_default",
    topic=KAFKA_TOPIC,
    producer_function=prod_function,
    producer_function_args=["{{ ti.xcom_pull(task_ids='get_number_of_treats')}}"],
    producer_function_kwargs={
        "pet_name": "{{ ti.xcom_pull(task_ids='get_your_pet_name')}}"
    },
    poll_timeout=10,
)
```

**Reply by potiuk** [reply]
Would be great to figure it out how to do it natively @mxmrlt  - maybe you would like to work on it ? There are enough OTEL enthusiasts amongh Airflow maintainers to help you succeed there :)
